packages/app-desktop/gui/NoteEditor/NoteBody/CodeMirror6/CodeMirror6.tsx

```diff
@@ -0,0 +1,593 @@
+
+// eslint-disable-next-line no-unused-vars
+import { EditorCommand, NoteBodyEditorProps, NoteBodyEditorRef } from '../../utils/types';
+import { getResourcesFromPasteEvent } from '../../utils/resourceHandling';
+import { ScrollOptions, ScrollOptionTypes } from '../../utils/types';
+import Toolbar from './Toolbar';
+// import styles_ from './styles';
+// import { RenderedBody, defaultRenderedBody } from './utils/types';
+import NoteTextViewer from '../../../NoteTextViewer';
+import Editor from './Editor';
+import usePluginServiceRegistration from '../../utils/usePluginServiceRegistration';
+import Setting from '@joplin/lib/models/Setting';
+import Note from '@joplin/lib/models/Note';
+import { _ } from '@joplin/lib/locale';
+import bridge from '../../../../services/bridge';
+// import markdownUtils from '@joplin/lib/markdownUtils';
+import shim from '@joplin/lib/shim';
+import { MenuItemLocation } from '@joplin/lib/services/plugins/api/types';
+import MenuUtils from '@joplin/lib/services/commands/MenuUtils';
+import CommandService from '@joplin/lib/services/CommandService';
+import SpellCheckerService from '@joplin/lib/services/spellChecker/SpellCheckerService';
+import convertToScreenCoordinates from '../../../utils/convertToScreenCoordinates';
+import { MarkupToHtml } from '@joplin/renderer';
+const { clipboard } = require('electron');
+const debounce = require('debounce');
+import shared from '@joplin/lib/components/shared/note-screen-shared';
+const Menu = bridge().Menu;
+const MenuItem = bridge().MenuItem;
+import { reg } from '@joplin/lib/registry';
+import ErrorBoundary from '../../../ErrorBoundary';
+import { MarkupToHtmlOptions } from '../../utils/useMarkupToHtml';
+import eventManager from '@joplin/lib/eventManager';
+import { EditContextMenuFilterObject } from '@joplin/lib/services/plugins/api/JoplinWorkspace';
+import type { ContextMenuEvent, ContextMenuParams } from 'electron';
+import usePrevious from '../../../hooks/usePrevious';
+import { EditorLanguageType, EditorSettings } from '@joplin/editor/types';
+import useStyles from './useStyles';
+import { EditorEvent, EditorEventType } from '@joplin/editor/events';
+import useScrollHandler from './useScrollHandler';
+import Logger from '@joplin/utils/Logger';
+import useEditorCommands from './useEditorCommands';
+import CodeMirrorControl from '@joplin/editor/CodeMirror/CodeMirrorControl';
+
+const menuUtils = new MenuUtils(CommandService.instance());
+
+const logger = Logger.create('CodeMirror6');
+const logDebug = (message: string) => logger.debug(message);
+
+interface RenderedBody {
+	html: string;
+	pluginAssets: any[];
+}
+
+function defaultRenderedBody(): RenderedBody {
+	return {
+		html: '',
+		pluginAssets: [],
+	};
+}
+
+function markupRenderOptions(override: MarkupToHtmlOptions = null): MarkupToHtmlOptions {
+	return { ...override };
+}
+
+function CodeMirror(props: NoteBodyEditorProps, ref: ForwardedRef<NoteBodyEditorRef>) {
```

Comment:
At present, `CodeMirror6.tsx` is mostly copied from `CodeMirror.tsx` for the CodeMirror 5 editor. A large amount of the code is the same.

Before merging, this pull request should perhaps attempt to break `CodeMirror.tsx` up into hooks/functions that can be shared by both the CodeMirror 5 and CodeMirror 6 setup.